src/dlfcn.c

```diff
@@ -467,7 +465,162 @@ char *dlerror( void )
+{
+    for( int i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++ ) {
+        PIMAGE_THUNK_DATA thunkILT = (PIMAGE_THUNK_DATA)( iid[i].Characteristics + baseAddress );
+        PIMAGE_THUNK_DATA thunkIAT = (PIMAGE_THUNK_DATA)( iid[i].FirstThunk + baseAddress );
+        for( ; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++ ) {
+            if( IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal) )
+                continue;
+            if( (void *)thunkIAT->u1.Function != addr )
+                continue;
+            PIMAGE_IMPORT_BY_NAME nameData = (PIMAGE_IMPORT_BY_NAME)( thunkILT->u1.AddressOfData + baseAddress );
+            return nameData->Name;
+        }
+   }
+   return NULL;
+}
+
+/*
+ * Return adress from Image Allocation Table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static unsigned char *getAddressFromIAT( unsigned char *addr )
+{
+    BYTE *p = (unsigned char *)addr;
+    /* ...inline app code...
+     * 00401002  |. E8 7B0D0000    CALL 00401D82               ; \GetModuleHandleA
+     * ...thunk table...
+     * 00401D82   $-FF25 4C204000 , JMP DWORD PTR DS:[40204C]  ;  KERNEL32.GetModuleHandleA
+     * ...memory address value of pointer...
+     * 40204C > FC 3D 57 7C   ;little endian pointer value
+     */
+    if( *p == 0xff && *(p+1) == 0x25 )
+    {
+        HMODULE module = GetModuleHandle( 0 );
+        DWORD size;
+        BYTE* iat = (BYTE *)getImportTable( module, &size );
+        if (!iat)
+            return NULL;
+#ifdef _WIN64
+        /* 0000000000401730 <dlsym>:
+         *    401730:	ff 25 ba 8c 00 00    	jmpq   *0x8cba(%rip)        # 40a3f0 <__imp_dlsym>
+         * 000000000040a3f0 <__imp_dlsym>:
+         *    40a3f0:	78 a7
+         *    40a3f2:	00 00
+         *    40a3f4:	00 00
+         */
+        ULONG offset = *(ULONG*)(p+2);
+        BYTE *p0 = (void *)(p + 6 + offset);
+        BYTE **p1 = p0;
+        //fprintf(stderr, "addr %p -> p0 %p p1 %p *p1 %p iat start %p end %p\n", addr, p0, p1, *p1, iat, iat+size);
+        if ( p1 < iat || p1 > iat + size )
+            return NULL;
+        //fprintf(stderr, "%p -> %p\n", addr, *p1);
+        return *p1;
+#else
+        BYTE **p1 = (void *)( p+2 );
+        //fprintf(stderr, "addr %p -> p1 %p *p1 %p iat start %p end %p\n", addr, p1, *p1, iat, iat+size);
+        if ( *p1 < iat || *p1 > iat + size )
+            return NULL;
+        BYTE **p2 = (void *)*p1;
+        //fprintf(stderr, "%p -> %p\n", addr, *p2);
+        return *p2;
+#endif
+    }
+    return NULL;
+}
+
+/* holds module filename */
+static char _module_filename[2*MAX_PATH];
+
+/**
+ * Get module information (filename and base address) from the address given
+ * @param addr address to get module info for
+ * @param info pointer to store module info
+ * @return 0 requested info filled into structure pointed by parameter info
+ * @return 1 error
+ */
+static int getModuleInfo( const void *addr, Dl_info *info )

```

Comment:
You can use `BOOL` type, `TRUE` for success and `FALSE` for error, like `GetModuleHandleExA()` or `DllMain()`.